File: 81.search-in-rotated-sorted-array-ii.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution81
{
    public bool Search(int[] nums, int target) {
        if(nums == null || nums.Length == 0 ) return false;

        return BinarySearch(nums, target, 0, nums.Length -1)
    }

    public bool BinarySearch(int[] nums, int target, int left, int right)
    {
        if(left > right) return false;
        if(left == right) return nums[left] == target;

        int middle = left + (right  - left) / 2;
        if(nums[middle] == target)
        {
            return true;
        }
        else if(nums[middle] < target)
        {
            return BinarySearch(nums, target, middle + 1, right) ||
                    (nums[left] >= nums[right]  && BinarySearch(nums, target, left, middle - 1));
        }
        else
        {
            return BinarySearch(nums, target, left, middle - 1) ||
                   (nums[left] >= nums[right]  && BinarySearch(nums, target, middle + 1, right));
        }
    }
}

File: 81_search-in-rotated-sorted-array-ii.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        if not nums or len(nums) == 0 : return False

        return self.binarySearch(nums, target, 0, len(nums) -1 )
    def binarySearch(self, nums, target, left, right) :
        if right < left : return False
        if right == left : return nums[left] == target
        middle = (left + right) // 2
        if nums[middle] == target: return True
        elif nums[middle] < target:
            return self.binarySearch(nums, target, middle + 1, right) or (nums[left] >= nums[right] and self.binarySearch(nums, target, left, middle -1))
        else:
            return self.binarySearch(nums, target, left, middle - 1) or (nums[left] >= nums[right] and self.binarySearch(nums, target, middle + 1, right))
        